arangod/Transaction/Helpers.cpp

```diff
@@ -433,6 +443,469 @@ std::string transaction::helpers::makeIdFromParts(
+    batchOptions.ensureComputedValuesContext(trx, collection);
+    batchOptions.computedValues->mergeComputedAttributes(
+        *batchOptions.computedValuesContext, trx, b->slice(), keysWritten,
+        ComputeValuesOn::kInsert, builder);
+  } else {
+    // add document as is
+    builder.add(b->slice());
+  }
+
+  TRI_ASSERT(revisionId.isSet());
+  return {};
+}
+
+/// @brief new object for replace, oldValue must have _key and _id correctly
+/// set
+Result transaction::helpers::newObjectForReplace(
+    transaction::Methods& trx, LogicalCollection& collection,
+    VPackSlice oldValue, VPackSlice newValue, RevisionId& revisionId,
+    VPackBuilder& builder, OperationOptions const& options,
+    transaction::BatchOptions& batchOptions) {
+  transaction::BuilderLeaser b(&trx);
+  b->openObject();
+
+  // add system attributes first, in this order:
+  // _key, _id, _from, _to, _rev
+
+  // _key
+  VPackSlice s = oldValue.get(StaticStrings::KeyString);
+  TRI_ASSERT(!s.isNone());
+  b->add(StaticStrings::KeyString, s);
+
+  // _id
+  s = oldValue.get(StaticStrings::IdString);
+  TRI_ASSERT(!s.isNone());
+  b->add(StaticStrings::IdString, s);
+
+  // _from and _to
+  if (collection.type() == TRI_COL_TYPE_EDGE) {
+    bool extendedNames = trx.vocbase().server().hasFeature<DatabaseFeature>() &&
+                         trx.vocbase()
+                             .server()
+                             .getFeature<DatabaseFeature>()
+                             .extendedNamesForCollections();
+
+    VPackSlice fromSlice = newValue.get(StaticStrings::FromString);
+    if (!isValidEdgeAttribute(fromSlice, extendedNames)) {
+      return Result(TRI_ERROR_ARANGO_INVALID_EDGE_ATTRIBUTE);
+    }
+
+    VPackSlice toSlice = newValue.get(StaticStrings::ToString);
+    if (!isValidEdgeAttribute(toSlice, extendedNames)) {
+      return Result(TRI_ERROR_ARANGO_INVALID_EDGE_ATTRIBUTE);
+    }
+
+    TRI_ASSERT(fromSlice.isString());
+    TRI_ASSERT(toSlice.isString());
+    b->add(StaticStrings::FromString, fromSlice);
+    b->add(StaticStrings::ToString, toSlice);
+  }
+
+  // _rev
+  bool handled = false;
+  if (options.isRestore) {
+    // copy revision id verbatim
+    s = newValue.get(StaticStrings::RevString);
+    if (s.isString()) {
+      b->add(StaticStrings::RevString, s);
+      VPackValueLength l;
+      char const* p = s.getStringUnchecked(l);
+      revisionId = RevisionId::fromString(p, l, false);
+      handled = true;
+    }
+  }
+  if (!handled) {
+    // temporary buffer for stringifying revision ids
+    char ridBuffer[arangodb::basics::maxUInt64StringSize];
+    revisionId = collection.newRevisionId();
+    b->add(StaticStrings::RevString, revisionId.toValuePair(&ridBuffer[0]));
+  }
+
+  containers::FlatHashSet<std::string_view> keysWritten;
+
+  // add other attributes after the system attributes
+  VPackObjectIterator it(newValue, true);
+  while (it.valid()) {
+    std::string_view key(it.key().stringView());
```

Comment:
```suggestion
    auto key = it.key().stringView();
```
suggest grep for that

Reply:
> Why not?

A VPackSlice only contains one member, which is a pointer. It can always be passed by value.
There is no need to hand it around or access it by const reference.